milli/src/update/index_documents/transform.rs

```diff
@@ -151,204 +126,99 @@ impl Transform<'_, '_> {
             self.max_memory,
         );
 
-        let mut json_buffer = Vec::new();
         let mut obkv_buffer = Vec::new();
-        let mut uuid_buffer = [0; uuid::adapter::Hyphenated::LENGTH];
         let mut documents_count = 0;
-
-        for result in documents {
-            let document = result.map_err(UserError::SerdeJson)?;
-
+        while let Some((addition_index, document)) = reader.next_document_with_index()? {
             if self.log_every_n.map_or(false, |len| documents_count % len == 0) {
-                progress_callback(UpdateIndexingStep::TransformFromUserIntoGenericFormat {
+                progress_callback(UpdateIndexingStep::RemapDocumentAddition {
                     documents_seen: documents_count,
                 });
             }
 
-            obkv_buffer.clear();
-            let mut writer = obkv::KvWriter::<_, FieldId>::new(&mut obkv_buffer);
+            let mut external_id_buffer = SmallVec::<[u8; 512]>::new();
+            let mut field_buffer = SmallVec::<[(u16, &[u8]); 128]>::new();
```

Comment:
I would like to get rid of these `SmallVec` things and use a normal `Vec` here. Could you also remove the `write` feature on the samllvec in the Cargo.toml too?

Reply:
So you would rather allocate all the time? what's the rationale?

Reply:
Allocating is OK here, that's a lot of boilerplate, maybe we can just reuse the `Vec<u8>` buffer by `clear`ing it like before. Adding smallvec just for this is too much boilerplate.

Reply:
No, this is why I used smallvec here, rust thinks that I'm keeping references in my vecs across loop iterations

Reply:
So, if you really want to keep these `SmallVec`s, please declare type aliases like `SmallVec512<T>` in the lib.rs. But I maintain that it is not a good thing to put 512 + (2 + 8) * 128 = 1 792 bytes on the stack.

Reply:
I would prefer that you declare a Vec buffer above the loop (131L) and clear it before we use it, like for the `obkv_buffer` before.

Reply:
You are right for the `external_id_buffer`. The issue with the `field_buffer` is that it holds references. I am not super happy with this solution either. I have made some ressearches:
- The first solution is to use unsafe code, and transmute the Vec to make the lifetimes match (after making sure that the vector is empty)
- We can use this crate: https://docs.rs/vec-utils/0.3.0/vec_utils/trait.VecExt.html#tymethod.drop_and_reuse to do it safely.

Reply:
@Kerollmops I made a new commit with the proposed solution, tell me what you think :)

Reply:
I must admit that I largely prefer this solution, even if importing a new dependency makes me feel bad :feelsgood:, at least it doesn't have any other dependency itself.

The problem is due to the `(u16, &[u8])` type keeping the lifetime metadata, I tried declaring an `fn move_me<T>(vec) -> Vec<T>` function but it wasn't erasing the lifetime meta as the `T` type wasn't altered, [the `drop_and_reuse` method](https://docs.rs/vec-utils/0.3.0/vec_utils/trait.VecExt.html#tymethod.drop_and_reuse) is changing from a `T` to a `U` type, that helps the compiler erase the lifetime.

Reply:
@Kerollmops, according to https://users.rust-lang.org/t/pattern-how-to-reuse-a-vec-str-across-loop-iterations/61657/4, you can do that:

```rust
fn reuse_vec<T, U>(mut v: Vec<T>) -> Vec<U> {
    assert_eq!(std::mem::size_of::<T>(), std::mem::size_of::<U>());
    assert_eq!(std::mem::align_of::<T>(), std::mem::align_of::<U>());
    v.clear();
    v.into_iter().map(|_| unreachable!()).collect()
}
```

which only works because of the optimization done in this pr https://github.com/rust-lang/rust/pull/70793

now if that's not black magic? haha

We can just use this function, but since it's rather confusing why it works, maybe we can keep the dependency. It's up to you.

Reply:
That's interesting but, as you say, too much black magic. We can keep this crate but please move it in the right place in the Cargo.toml dependencies 😃 What a story, I tried for maybe 30mins to make it work without `unsafe` and any external dependencies!